core/src/test/java/org/elasticsearch/index/IndexWithShadowReplicasIT.java

```diff
@@ -793,35 +806,146 @@ public void testDeletingClosedIndexRemovesFiles() throws Exception {
+        flushAndRefresh(IDX);
 
         logger.info("--> closing index {}", IDX);
         client().admin().indices().prepareClose(IDX).get();
 
-        logger.info("--> deleting non-closed index");
-        client().admin().indices().prepareDelete(IDX2).get();
-        assertPathHasBeenCleared(dataPath2);
         logger.info("--> deleting closed index");
         client().admin().indices().prepareDelete(IDX).get();
+
+        assertPathHasBeenCleared(dataPath);
+        assertIndicesDirsDeleted(nodes);
+    }
+
+    public void testDeletingIndexWithDedicatedMasterNodes() throws Exception {
+        final Path dataPath = createTempDir();
+        final Settings nodeSettings = nodeSettings(dataPath.getParent());
+        final List<String> nodes = internalCluster().startMasterOnlyNodesAsync(1, nodeSettings).get();
+        assert nodes.size() > 0;
+        logger.info("--> master node is " + nodes.get(0));
+        final int numDataNodes = randomIntBetween(2, 5);
+        nodes.addAll(internalCluster().startDataOnlyNodesAsync(numDataNodes, nodeSettings).get());
+        final String IDX = "test";
+        final Settings idxSettings = Settings.builder()
+                                             .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 5)
+                                             .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, numDataNodes - 1)
+                                             .put(IndexMetaData.SETTING_DATA_PATH, dataPath.toAbsolutePath().toString())
+                                             .put(IndexMetaData.SETTING_SHADOW_REPLICAS, true)
+                                             .put(IndexMetaData.SETTING_SHARED_FILESYSTEM, true)
+                                             .build();
+
+        prepareCreate(IDX).setSettings(idxSettings).addMapping("doc", "foo", "type=text").get();
+        ensureGreen(IDX);
+
+        client().prepareIndex(IDX, "doc", "1").setSource("foo", "bar").get();
+        client().prepareIndex(IDX, "doc", "2").setSource("foo", "bar").get();
+        flushAndRefresh(IDX);
+
+        GetResponse gResp1 = client().prepareGet(IDX, "doc", "1").setFields("foo").get();
+        GetResponse gResp2 = client().prepareGet(IDX, "doc", "2").setFields("foo").get();
+        assertThat(gResp1.getField("foo").getValue().toString(), equalTo("bar"));
+        assertThat(gResp2.getField("foo").getValue().toString(), equalTo("bar"));
+
+        logger.info("--> performing query");
+        SearchResponse resp = client().prepareSearch(IDX).setQuery(matchAllQuery()).get();
+        assertHitCount(resp, 2);
+
+        assertAcked(client().admin().indices().prepareDelete(IDX));
+
+        assertPathHasBeenCleared(dataPath);
+        //norelease
+        //TODO: uncomment the test below when https://github.com/elastic/elasticsearch/issues/17695 is resolved.
+        //assertIndicesDirsDeleted(nodes);
+    }
+
+    public void testShadowReplicaIndexWithNodesHavingNoShards() throws Exception {
```

Comment:
do we need a dedicated test for this? can't we randomize the number of nodes in the normal deletion tests? O.w. we will need to test this situation with both closed indices and open. It's a shame IMO